scripts/speech_recognition/estimate_duration_bins_2d.py

```diff
@@ -0,0 +1,319 @@
+
+
+def estimate_duration_buckets(
+    cuts: Iterable[Cut],
+    num_buckets: int,
+    num_subbuckets: int,
+    max_tps: float,
+    max_duration: float,
+    quiet: bool,
+) -> list[tuple[float, float]]:
+    assert num_buckets > 1
+
+    constraint = FixedBucketBatchSizeConstraint2D([(0.0, 0.0)], [0])
+
+    sizes = []
+    num_tokens = []
+    for c in cuts:
+        dur, toks = constraint.measure_length(c)
+        sizes.append(dur)
+        num_tokens.append(toks)
+    sizes = np.array(sizes, dtype=np.float32)
+    num_tokens = np.array(num_tokens, dtype=np.int32)
+    joint = np.rec.fromarrays([sizes, num_tokens])
+    joint.sort()
+    sizes = joint.f0
+    num_tokens = joint.f1
+
+    size_per_bucket = sizes.sum() / num_buckets
+
+    if not quiet:
+        print("Duration distribution:")
+        print(pd.Series(sizes).describe(percentiles=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]))
+    if math.isinf(max_duration):
+        max_duration = sizes[-1]
+
+    tps = num_tokens / sizes
+    if not quiet:
+        print("Token per second distribution:")
+        print(pd.Series(tps).describe(percentiles=[0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99]))
+    if math.isinf(max_tps):
+        max_tps = tps.max()
+    del tps
+
+    bins = []
+    bin_indexes = [0]
+    tot = 0.0
+    for binidx, size in enumerate(sizes):
+        if tot > size_per_bucket:
+            num_tokens_bucket = num_tokens[bin_indexes[-1] : binidx]
+            num_tokens_bucket.sort()
+            tokens_per_subbucket = num_tokens_bucket.sum() / num_subbuckets
+            tot_toks = 0
+            for num_toks in num_tokens_bucket:
+                if tot_toks > tokens_per_subbucket:
+                    bins.append((size, num_toks))
+                    tot_toks = 0
+                tot_toks += num_toks
+            bins.append((size, math.ceil(size * max_tps)))
+            bin_indexes.append(binidx)
+            tot = 0.0
+        tot += size
+
+    num_tokens_bucket = num_tokens[bin_indexes[-1] : binidx]
+    num_tokens_bucket.sort()
+    tokens_per_subbucket = num_tokens_bucket.sum() / num_subbuckets
+    tot_toks = 0
+    for num_toks in num_tokens_bucket:
+        if tot_toks > tokens_per_subbucket:
+            bins.append((max_duration, num_toks))
+            tot_toks = 0
+        tot_toks += num_toks
+    bins.append((max_duration, math.ceil(max_duration * max_tps)))
+
+    return bins
+
+
+def load_tokenizer(paths: list[str], langs: list[str] = None) -> TokenizerWrapper:
+    if len(paths) == 1:
+        tok = SentencePieceTokenizer(paths[0])
+    else:
+        assert langs is not None and len(paths) == len(
+            langs
+        ), f"Cannot create AggregateTokenizer; each tokenizer must have assigned a language via --langs option (we got --tokenizers={paths} and --langs={langs})"
+        tok = AggregateTokenizer({lang: SentencePieceTokenizer(p) for lang, p in zip(langs, paths)})
+    return TokenizerWrapper(tok)
```

Comment:
A note: when working with speechLM models, sometimes we also need to support hugginface tokenizers (e.g. llama3.1). NeMo has a wrapper around hugging face tokenizer. https://github.com/NVIDIA/NeMo/blob/9f9a01affd16bb6a4f730cb312c73d728bbb2877/nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py#L28

Reply:
Good point. Let's add the support for that in a follow up PR (or speechlm branch) once this is merged.